### PR TITLE
ec2: fix instids and res_list being referenced before assigned - fixes #22692

### LINKED ISSUE #22692: Creating EC2 spot requests without waiting for instances to start results in a Python error
_From @kyl191 on September 24, 2015 2:59_

##### Ansible Version: 1.9.3

```
[root@ansible-host ansible]# ansible --version
ansible 1.9.3
  configured module search path = None
```
##### Ansible Configuration:

hostfile was changed
##### Environment:

N/A
##### Summary:

Using the EC2 module, creating a spot but not waiting for the spot request to be fulfilled results in a Python error
##### Steps To Reproduce:
1. Use this role file:

``` yml
    - name: Create spot instance
      local_action:
        module: ec2
        spot_price: 0.24
        spot_wait_timeout: 600
        region: ap-southeast-1
        keypair: EC2 AWS APAC Singapore
        group_id: default
        instance_type: c3.large
        image: ami-ac9481fe
        wait: no
```
1. Run ansible against role file
##### Expected Results:

An AWS spot request is submitted, and the playbook finishes
##### Actual Results:

An AWS spot request is submitted, and Python crashes with an error:

```
    Traceback (most recent call last):
      File "/root/.ansible/tmp/ansible-tmp-1443062264.07-30079051580954/ec2", line 3070, in <module>
        main()
      File "/root/.ansible/tmp/ansible-tmp-1443062264.07-30079051580954/ec2", line 1249, in main
        (instance_dict_array, new_instance_ids, changed) = create_instances(module, ec2, vpc)
      File "/root/.ansible/tmp/ansible-tmp-1443062264.07-30079051580954/ec2", line 982, in create_instances
        while wait_timeout > time.time() and num_running < len(instids):
    UnboundLocalError: local variable 'instids' referenced before assignment
```

`instids` is created inside if/else blocks, and creating a spot instance but not waiting isn't one of the conditions checked. 
Adding an else block to the if block here https://github.com/ansible/ansible-modules-core/blob/devel/cloud/amazon/ec2.py#L1039 should work, something like

``` python
else:
  instids = None
```

_Copied from original issue: ansible/ansible-modules-core#2140_

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
Fixes #22692.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2.py

##### ANSIBLE VERSION
```
ansible 2.4.0 (ec2-spot-instance-fix a099c6e310) last updated 2017/03/28 15:56:19 (GMT -400)
  config file =
  configured module search path = Default w/o overrides
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2.py [hunk 1]
```diff
@@ -1227,12 +1227,15 @@ def create_instances(module, ec2, vpc, override_count=None):
...
                     instids = await_spot_requests(module, ec2, res, count)
+                else:
+                    instids = []
         except boto.exception.BotoServerError as e:
             module.fail_json(msg="Instance creation failed => %s: %s" % (e.error_code, e.error_message))
 
         # wait here until the instances are up
         num_running = 0
         wait_timeout = time.time() + wait_timeout
+        res_list = None
```
Comment (abadger) [inline lib/ansible/modules/cloud/amazon/ec2.py]: Look like you could set
``` python
res_list = tuple()
```
here and then the for loops can just be:
``` python
for res in res_list:
```

File: lib/ansible/modules/cloud/amazon/ec2.py [hunk 2]
```diff
@@ -1227,12 +1227,15 @@ def create_instances(module, ec2, vpc, override_count=None):
                 # Now we have to do the intermediate waiting
                 if wait:
                     instids = await_spot_requests(module, ec2, res, count)
+                else:
+                    instids = []
```
Comment (abadger) [inline lib/ansible/modules/cloud/amazon/ec2.py:1231]: Do we also need to guard the code here: https://github.com/ansible/ansible/pull/23053/files#diff-dea76a894873bd8fbd3173b3a8610a11L1280 or does  ec2.create_tags() handle an empty list fine?

**Comment by abadger** [comment]
Looks pretty straightforward.  Left a couple questions but otherwise looks fine.   Feel free to merge once those are addressed/tested.

File: lib/ansible/modules/cloud/amazon/ec2.py (hunk 2 above)
Comment (willthames) [reply lib/ansible/modules/cloud/amazon/ec2.py:1231]: Looks like we do need to guard against an empty `instids` list:

```
----> 1 ec2.create_tags([],{})

/usr/lib/python2.7/site-packages/boto/ec2/connection.pyc in create_tags(self, resource_ids, tags, dry_run)
   4217         if dry_run:
   4218             params['DryRun'] = 'true'
-> 4219         return self.get_status('CreateTags', params, verb='POST')
   4220 
   4221     def delete_tags(self, resource_ids, tags, dry_run=False):

/usr/lib/python2.7/site-packages/boto/connection.pyc in get_status(self, action, params, path, parent, verb)
   1225             boto.log.error('%s %s' % (response.status, response.reason))
   1226             boto.log.error('%s' % body)
-> 1227             raise self.ResponseError(response.status, response.reason, body)

EC2ResponseError: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>MissingParameter</Code><Message>The request must contain the parameter resourceIdSet</Message></Error></Errors><RequestID>62f52f6f-c907-412a-a4d0-68ce1f54acc2</RequestID></Response>
```

Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/ec2.py:1231]: @willthames, doesn't my commit address it?

**Comment by willthames** [comment]
shipit

**Comment by jucor** [comment]
Yay, thanks!

**Pull request merged**